src/router.js

```diff
@@ -0,0 +1,290 @@
+    if (permissionProvider && typeof permissionProvider !== 'function') {
+      throw new TypeError('Permission provider must be a function');
+    }
+
+    this._permissionProvider = permissionProvider;
+
+    return this;
+  }
+
+  registerRoute(method, pattern, routeSchema, handler, appendToSchema) {
+    appendToSchema = typeof appendToSchema === 'undefined'
+      ? true
+      : appendToSchema;
+
+    if (this._isLoaded) {
+      throw new Error('Cannot add route to a loaded router.');
+    }
+
+    const route = {
+      method: method.toLowerCase(),
+      pattern: pattern,
+      schema: routeSchema || {},
+    };
+
+    const routeKey = `${method}:${pattern}`;
+
+    const schemaSpec = operationSchema.validate(route.schema);
+    const error = [];
+    if (schemaSpec.error) {
+      error.push(schemaSpec.error);
+    }
+    if (this._routes[routeKey]){
+      error.push(`Duplicate route ${routeKey}`);
+    }
+    if (!this._expressApp[route.method] || typeof this._expressApp[route.method] !== 'function') {
+      error.push(`Incorrect method ${route.method}`);
+    }
+
+    if (error.length) {
+      route.errors = error;
+      return this._events.emit(Router.REGISTRATION_ERROR, route);
+    }
+
+    route.schema = schemaSpec.value;
+    handler = handler || route.schema.handler;
+    const handlers = [];
+
+    if (route.schema) {
+      if (route.schema.security && Object.keys(route.schema.security).length) {
+        // Secured route - add security check
+        handlers.push(
+          (req, res, next) => {
+            securityValidation(
+                this._permissionProvider,
+                route.schema.security,
+                req)
+              .then(() => {
+                next();
+              })
+              .catch((err) => {
+                return next(new HttpError.Unauthorized(JSON.stringify(err)));
+              });
+          });
+      }
+
+      if (route.schema.parameters && route.schema.parameters.length) {
+        // Has parameters - add parameter check
+        handlers.push(
+          (req, res, next) => {
+            const errors = {};
+
+            try {
+              route.schema.parameters
+                .forEach((param) => {
+                  const result = parameterValidation(param, req);
+
+                  if (result !== true) {
+                    return next(new HttpError.BadRequest(JSON.stringify(result)));
+                  }
+
+                  return next();
+                });
+            }
+            catch (err) {
+              console.error(err);
+              return next(new HttpError.InternalServerError());
+            }
+          }
+        )
+      }
+    }
+
+    if (handler && typeof handler === 'function') {
+      // Actual route handler
+      handlers.push(handler);
+    }
+    else {
+      // Default handler - not sure if we want this
+      handlers.push((res, req, next) => {
+        res.status(200).send();
```

Comment:
A `501 Not Implemented` might make the most sense at this point, keeps it pretty opinion free